File: src/main/java/dev/ssdd/ZotPrivateKey.java
package dev.ssdd;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

import static dev.ssdd.BigIntPlayground.*;

public class ZotPrivateKey {

    private final BigInteger on; // O(n)
    private final BigInteger d // d
    ;
    public ZotPrivateKey(int bitLen, int nx, ZotKeyPair zotKeyPair, Encrypto encrypto) {
        SecureRandom secureRandom = new SecureRandom();
        BigInteger p = new BigInteger(bitLen, nx, secureRandom);//prime no p
        BigInteger q = new BigInteger(bitLen, nx, secureRandom);//prime no q

        // p*q
        BigInteger n = p.multiply(q);
        // p-1
        BigInteger pm1 = p.subtract(BigInteger.ONE);
        // q-1
        BigInteger qm1 = q.subtract(BigInteger.ONE);
        on = pm1.multiply(qm1);
        // e
        BigInteger e = new BigInteger(bitLen, nx, new SecureRandom());
        d = e.modInverse(on); //d*e mod O(n)

        try{
            assertEquals(e.multiply(d).mod(on));
            zotKeyPair.publicKey = new ZotPublicKey(e, n);
            encrypto.keyPair = zotKeyPair;
        }catch (AssertionError er){
            er.printStackTrace();
        }
    }

    private void assertEquals(BigInteger mod) throws AssertionError{
        if (!BigInteger.ONE.equals(mod)) throw new AssertionError("defect in generating keys.");
    }

    public String decrypt(byte[] decode) {
        return new String(writeBigInt((convertBytesToBigInt(decode).multiply(d)).mod(on)), StandardCharsets.UTF_8);
    }

    @Override
    public String toString() {
        return "Algo: RSA\n" + "modulus: "+on + "\nprivate exponent: "+d;
    }
}

File: lib/zot_private_key.dart
import 'dart:convert';

import 'package:encrypto_flutter/encrypto_flutter.dart';
import 'package:encrypto_flutter/keypair.dart';
import 'package:encrypto_flutter/primegen.dart';
import 'package:encrypto_flutter/zot_public_key.dart';

class ZotPrivateKey {
  late BigInt _on, // modulus (O(n))
      _d; // private exponent
  ZotPrivateKey(int bitLen, ZotKeyPair zotKeyPair, Encrypto encrypto) {
    BigInt p = randomPrimeBigInt(bitLen), q = randomPrimeBigInt(bitLen);
    BigInt n = p * q, pm1 = p - BigInt.one, qm1 = q - BigInt.one;
    _on = pm1 * qm1;
    BigInt e = randomPrimeBigInt(bitLen);
    _d = e.modInverse(_on);
    _assetEq(((e * _d) % _on));
    zotKeyPair.publicKey = ZotPublicKey(e, n);
    encrypto.keyPair = zotKeyPair;
  }

  /// returns decrypted message
  String decrypt(String val) {
    return String.fromCharCodes(ZotPublicKey.writeBigInt(
        (ZotPublicKey.readBytes(const Base64Decoder().convert(val)) * _d) %
            _on));
  }

  /// returns info about private key: the information contains:
  /// 1. algo, 2. modulus, 3. private exponent
  @override
  String toString() {
    // TODO: implement toString
    return "Algo: RSA\nmodulus: ${_on.toString()}\nprivate exponent: ${_d.toString()}";
  }

  /// to make sure generated numbers works properly.
  void _assetEq(BigInt bigInt) {
    if (bigInt.toInt() != 1) {
      throw Exception("defect in generating keys.");
    }
  }
}
